server/sonar-server/src/main/java/org/sonar/server/authentication/GenerateJwtTokenFilter.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.authentication;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.WriteListener;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import org.sonar.api.web.ServletFilter;
+import org.sonar.server.user.UserSession;
+
+import static org.jboss.netty.handler.codec.http.HttpMethod.POST;
+
+/**
+ * This filter should only be executed on "/sessions/login".
+ *
+ * It will create a new session when user is authenticated.
+ */
+public class GenerateJwtTokenFilter extends ServletFilter {
+
+  private final JwtTokenUpdater jwtTokenUpdater;
+  private final UserSession userSession;
+
+  public GenerateJwtTokenFilter(JwtTokenUpdater jwtTokenUpdater, UserSession userSession) {
+    this.jwtTokenUpdater = jwtTokenUpdater;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public UrlPattern doGetPattern() {
+    return UrlPattern.create("/sessions/login");
+  }
+
+  @Override
+  public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain) throws IOException, ServletException {
+    HttpServletRequest request = (HttpServletRequest) servletRequest;
+    HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+    if (request.getMethod().equals(POST.getName())) {
+      BufferResponseWrapper wrapper = new BufferResponseWrapper(response);
+      chain.doFilter(request, wrapper);
+      if (userSession.isLoggedIn()) {
+        jwtTokenUpdater.createNewJwtToken(userSession.getLogin(), response);
+      }
```

Comment:
After offline discussion, we won't do that because otherwise it will remove too much things, for instance the 'return_to' parameter that is for the moment stored in session.